File: 2025/javascript/src/day01.js
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const MODULUS = 100;
const __dirname = path.dirname(fileURLToPath(import.meta.url));
const projectRoot = path.resolve(__dirname, "..");
const defaultInputPath = path.resolve(projectRoot, "../inputs/day01.txt");

const mod = (value, divisor) => ((value % divisor) + divisor) % divisor; // keep remainder positive
const quot = (value, divisor) => Math.trunc(value / divisor);

export const parseStep = (line) => {
  const magnitude = Number(line.slice(1));

  if (Number.isNaN(magnitude)) {
    throw new Error(`Invalid step: ${line}`);
  }

  return line.startsWith("L") ? -magnitude : magnitude;
};

export const loadSteps = (inputPath = defaultInputPath) => {
  const resolvedPath = path.isAbsolute(inputPath)
    ? inputPath
    : path.resolve(projectRoot, inputPath);

  const lines = fs
    .readFileSync(resolvedPath, "utf8")
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter(Boolean);

  return { inputPath: resolvedPath, lines, steps: lines.map(parseStep) };
};

export const countReturnsToZero = (steps) => {
  const { reached0 } = steps.reduce(
    ({ dial, reached0: hits }, step) => {
      const dialAfterStep = dial + step;
      const hit = dialAfterStep % MODULUS === 0;

      return {
        dial: dialAfterStep,
        reached0: hit ? hits + 1 : hits,
      };
    },
    { dial: 50, reached0: 0 },
  );

  return reached0;
};

export const countReturnsToZeroCrossings = (steps) => {
  const { reached0 } = steps.reduce(
    ({ dial, reached0: hits }, step) => {
      const fullDials = Math.abs(quot(step, MODULUS));
      const stepRem = step % MODULUS;
      const modPlusDial = dial + stepRem;
      const crossed =
        dial !== 0 && (modPlusDial < 0 || modPlusDial > MODULUS);
      const dialWrapped = mod(dial + step, MODULUS);
      const landedOnMark = dialWrapped === 0;

      return {
        dial: dialWrapped,
        reached0:
          hits +
          fullDials +
          (crossed ? 1 : 0) +
          (landedOnMark ? 1 : 0),
      };
    },
    { dial: 50, reached0: 0 },
  );

  return reached0;
};

export const countReturnsToZeroCrossingsReduce = (steps) => {
  // Variant mirroring Day01TwoReduce from the Java version.
  const { reached0 } = steps.reduce(
    ({ position, reached0: hits }, nextDial) => {
      const fullDials = Math.abs(quot(nextDial, MODULUS));
      const stepRem = nextDial % MODULUS;
      const modPlusDial = position + stepRem;
      const crossed =
        position !== 0 && (modPlusDial < 0 || modPlusDial > MODULUS);
      const positionWrapped = mod(position + nextDial, MODULUS);
      const landedOnMark = positionWrapped === 0;

      return {
        position: positionWrapped,
        reached0:
          hits +
          fullDials +
          (crossed ? 1 : 0) +
          (landedOnMark ? 1 : 0),
      };
    },
    { position: 50, reached0: 0 },
  );

  return reached0;
};

export const solvePart1 = (inputPath) => {
  const { inputPath: resolved, lines, steps } = loadSteps(inputPath);
  return {
    inputPath: resolved,
    linesRead: lines.length,
    result: countReturnsToZero(steps),
  };
};

export const solvePart2 = (inputPath) => {
  const { inputPath: resolved, lines, steps } = loadSteps(inputPath);
  return {
    inputPath: resolved,
    linesRead: lines.length,
    result: countReturnsToZeroCrossings(steps),
  };
};

export const solvePart2Reduce = (inputPath) => {
  const { inputPath: resolved, lines, steps } = loadSteps(inputPath);
  return {
    inputPath: resolved,
    linesRead: lines.length,
    result: countReturnsToZeroCrossingsReduce(steps),
  };
};

File: 2025/python/src/day01.py
from __future__ import annotations

import math
from functools import reduce
from pathlib import Path
from typing import Iterable, Sequence, TypedDict

MODULUS = 100
START_POSITION = 50
PROJECT_ROOT = Path(__file__).resolve().parent.parent
DEFAULT_INPUT_PATH = (PROJECT_ROOT.parent / "inputs/day01.txt").resolve()


class LoadedSteps(TypedDict):
    input_path: Path
    lines: list[str]
    steps: list[int]


class SolveResult(TypedDict):
    input_path: Path
    lines_read: int
    result: int


def mod(value: int, divisor: int) -> int:
    return ((value % divisor) + divisor) % divisor


def quot(value: int, divisor: int) -> int:
    return math.trunc(value / divisor)


def remainder(value: int, divisor: int) -> int:
    # Match Java's remainder that keeps the sign of the dividend.
    return value - quot(value, divisor) * divisor


def parse_step(line: str) -> int:
    stripped = line.strip()
    if not stripped:
        raise ValueError("Empty step")

    magnitude = int(stripped[1:])
    return -magnitude if stripped.startswith("L") else magnitude


def load_steps(input_path: str | Path | None = None) -> LoadedSteps:
    resolved = Path(input_path) if input_path is not None else DEFAULT_INPUT_PATH
    if not resolved.is_absolute():
        resolved = (PROJECT_ROOT / resolved).resolve()

    lines = [line.strip() for line in resolved.read_text().splitlines() if line.strip()]
    return {"input_path": resolved, "lines": lines, "steps": [parse_step(line) for line in lines]}


def count_returns_to_zero(steps: Sequence[int]) -> int:
    dial = START_POSITION
    reached0 = 0

    for step in steps:
        dial += step
        if dial % MODULUS == 0:
            reached0 += 1

    return reached0


def count_returns_to_zero_crossings(steps: Sequence[int]) -> int:
    dial = START_POSITION
    reached0 = 0

    for step in steps:
        full_dials = abs(quot(step, MODULUS))
        step_rem = remainder(step, MODULUS)
        mod_plus_dial = dial + step_rem
        crossed = dial != 0 and (mod_plus_dial < 0 or mod_plus_dial > MODULUS)
        dial_wrapped = mod(dial + step, MODULUS)
        landed_on_mark = dial_wrapped == 0

        reached0 += full_dials + (1 if crossed else 0) + (1 if landed_on_mark else 0)
        dial = dial_wrapped

    return reached0


def count_returns_to_zero_crossings_reduce(steps: Sequence[int]) -> int:
    def reducer(state: tuple[int, int], next_dial: int) -> tuple[int, int]:
        position, hits = state

        full_dials = abs(quot(next_dial, MODULUS))
        step_rem = remainder(next_dial, MODULUS)
        mod_plus_dial = position + step_rem
        crossed = position != 0 and (mod_plus_dial < 0 or mod_plus_dial > MODULUS)
        position_wrapped = mod(position + next_dial, MODULUS)
        landed_on_mark = position_wrapped == 0

        reached = hits + full_dials + (1 if crossed else 0) + (1 if landed_on_mark else 0)
        return position_wrapped, reached

    _, reached0 = reduce(reducer, steps, (START_POSITION, 0))
    return reached0


def solve_part1(input_path: str | Path | None = None) -> SolveResult:
    loaded = load_steps(input_path)
    steps: Iterable[int] = loaded["steps"]
    return {
        "input_path": loaded["input_path"],
        "lines_read": len(loaded["lines"]),
        "result": count_returns_to_zero(steps),
    }


def solve_part2(input_path: str | Path | None = None) -> SolveResult:
    loaded = load_steps(input_path)
    steps: Iterable[int] = loaded["steps"]
    return {
        "input_path": loaded["input_path"],
        "lines_read": len(loaded["lines"]),
        "result": count_returns_to_zero_crossings(steps),
    }


def solve_part2_reduce(input_path: str | Path | None = None) -> SolveResult:
    loaded = load_steps(input_path)
    steps: Iterable[int] = loaded["steps"]
    return {
        "input_path": loaded["input_path"],
        "lines_read": len(loaded["lines"]),
        "result": count_returns_to_zero_crossings_reduce(steps),
    }
